Documentation/Assembler.md

```diff
@@ -0,0 +1,148 @@
+# Modularizing Service Registration
+This feature provides your implementation the ability to group related service definitions together
+in an `AssemblyType`. This allows your application to:
+
+   - keep things organized by keeping like services in 1 place
+   - provided a shared `Container`
+   - allows registering different assembly configurations which is useful for swapping out mock implementations
+   - can be "load aware" when the container is fully configured
+
+This feature is an opinionated way to how your can register services in your `Container`. There are
+parts to this feature:
+
+## AssemblyType
+The `AssemblyType` is a protocol that is provided a shared `Container` where service definitions 
+can be registered. The shared `Container` will contain **all** service definitions from every
+`AssemblyType` registered to the `Assembler`. Let's look at an example:
+
+    class ServiceAssembly: AssemblyType {
+        func assemble(container: Container) {
+            container.register(FooServiceType.self) { r in 
+               return FooService()
+            }
+            container.register(BarServiceType.self) { r in 
+               return BarService()
+            }
+        }
+    }
+
+    class ManagerAssembly: AssemblyType {
+        func assemble(container: Container) {
+            container.register(FooManagerType.self) { r in 
+               return FooManager(service: r.resolve(FooServiceType.self)!)
+            }
+            container.register(BarManagerType.self) { r in 
+               return BarManager(service: r.resolve(BarServiceType.self)!)
+            }
+        }
+    }
+
+Here we have created 2 assemblies: 1 for services and 1 for managers. As you can see the `ManagerAssembly`
+leverages service definitions registered in the `ServiceAssembly`. Using this pattern the `ManagerAssembly`
+doesn't care where the `FooServiceType` and `BarServiceType` are registered, it just requires them to
+be registered else where.
+
+### AssemblyLoadAwareType
```

Comment:
Yea I can make this change. I thought it would be better to be explicit about which assemblies are going to run code after they are loaded by looking at its protocol name, but I don't ming the no-op function.